### PR TITLE
Implement split()/merge() API of the Coarse provider

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Implement more API of the Coarse provider.

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

File: include/umf/providers/provider_coarse.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#ifndef UMF_COARSE_PROVIDER_H
+#define UMF_COARSE_PROVIDER_H
+
+#include <stdbool.h>
+#include <umf/memory_provider.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
```
Comment (bratpiorka) [inline include/umf/providers/provider_coarse.h:17]: add "/// @brief Coarse Memory Provider allocation strategy"

Comment (ldorau) [reply include/umf/providers/provider_coarse.h:17]: Done

File: src/provider/provider_coarse.c [hunk 2]
```diff
@@ -1517,6 +1517,179 @@ coarse_memory_provider_get_stats(void *provider,
...
+    block->size = firstSize;
+    new_block->used = true;
+
+    assert(new_block->size == (totalSize - firstSize));
+
+    assert(debug_check(coarse_provider));
+
+    if (utils_mutex_unlock(&coarse_provider->lock) != 0) {
+        LOG_ERR("unlockng the lock failed");
+        return UMF_RESULT_ERROR_UNKNOWN;
```
Comment (bratpiorka) [inline src/provider/provider_coarse.c:1579]: NIT: you can change the flow slightly to a simpler one - initializing umf_result at the beginning and calling unlock only once

**Review by bratpiorka — APPROVED** [review]

File: src/provider/provider_coarse.c (hunk 2 above)
Comment (ldorau) [reply src/provider/provider_coarse.c]: Done

Comment (ldorau) [reply src/provider/provider_coarse.c]: Done in the next function too.

File: src/provider/provider_coarse.c [hunk 3]
```diff
@@ -1517,6 +1517,179 @@ coarse_memory_provider_get_stats(void *provider,
...
+    assert(merged_node == low_node);
+    assert(low_block->size == totalSize);
+
+    umf_result = UMF_RESULT_SUCCESS;
+
+err_mutex_unlock:
+    assert(debug_check(coarse_provider));
+
+    if (utils_mutex_unlock(&coarse_provider->lock) != 0) {
+        LOG_ERR("unlockng the lock failed");
```
Comment (lukaszstolarczuk) [inline src/provider/provider_coarse.c:1684]: .

File: test/disjointCoarseMallocPool.cpp [hunk 4]
```diff
@@ -759,3 +759,95 @@ TEST_P(CoarseWithMemoryStrategyTest, disjointCoarseMallocPool_wrong_params_4) {
 
     umfMemoryProviderDestroy(malloc_memory_provider);
 }
+
+TEST_P(CoarseWithMemoryStrategyTest, disjointCoarseMallocPool_split_merge) {
```
Comment (lukaszstolarczuk) [inline test/disjointCoarseMallocPool.cpp:763]: would be nice to have some negative tests as well

File: src/provider/provider_coarse.c (hunk 3 above)
Comment (ldorau) [reply src/provider/provider_coarse.c]: Done

File: src/provider/provider_coarse.c [hunk 5]
```diff
@@ -1517,6 +1517,179 @@ coarse_memory_provider_get_stats(void *provider,
...
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_result_t umf_result;
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    if (utils_mutex_lock(&coarse_provider->lock) != 0) {
+        LOG_ERR("lockng the lock failed");
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_coarse.c]: Done

File: src/provider/provider_coarse.c [hunk 6]
```diff
@@ -1517,6 +1517,179 @@ coarse_memory_provider_get_stats(void *provider,
...
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_coarse.c]: Done

**Comment by ldorau** [comment]
@lukaszstolarczuk all issues fixed

File: src/provider/provider_coarse.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_coarse.c]: @lukaszstolarczuk ping

File: src/provider/provider_coarse.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_coarse.c]: @lukaszstolarczuk ping

File: src/provider/provider_coarse.c (hunk 3 above)
Comment (ldorau) [reply src/provider/provider_coarse.c]: @lukaszstolarczuk ping

File: test/disjointCoarseMallocPool.cpp (hunk 4 above)
Comment (ldorau) [reply test/disjointCoarseMallocPool.cpp:763]: @lukaszstolarczuk ping

Comment (lukaszstolarczuk) [reply test/disjointCoarseMallocPool.cpp:763]: perfect, thank you!

**Review by lukaszstolarczuk — COMMENTED** [review]
LGTM

**Review by lplewa — APPROVED** [review]

**Pull request merged**